File: java/leetcode/backtracking/combinations.java
/**
 * 77. Combinations
 * Given two integers n and k, return all possible combinations of k numbers out of 1 ... n.

Example:

Input: n = 4, k = 2
Output:
[
  [2,4],
  [3,4],
  [2,3],
  [1,2],
  [1,3],
  [1,4],
]
 */

import java.util.List;
import java.util.ArrayList;

public class combinations {

    public List<List<Integer>> combine(int n, int k) {
        List<List<Integer>> combinations = new ArrayList<>();
        List<Integer> combineList = new ArrayList<>();
        backtracking(combineList, combinations, 1, k, n);
        return combinations;
    }

    private void backtracking(List<Integer> combineList, List<List<Integer>> combinations, 
        int start, int k, final int n) {
    
        if (k == 0) {
            combinations.add(new ArrayList<>(combineList));
            return;
        }
        for (int i = start; i <= n-k+1; i++) {
            combineList.add(i);
            backtracking(combineList, combinations, i+1, k-1, n);
            combineList.remove(combineList.size()-1);
        }
    }
}

File: golang/leetcode/backtracking/combinations.go
package backtracking

// 77. Combinations

func combine(n int, k int) [][]int {
	combinations := make([][]int, 0)
	temp := make([]int, 0)
	combineTrack(temp, &combinations, 1, k, n)
	return combinations
}

func combineTrack(temp []int, combinations *[][]int, start int, k int, n int) {
	if k == 0 {
		combo := make([]int, len(temp))
		copy(combo, temp)
		*combinations = append(*combinations, combo)
		return
	}
	for i := start; i <= n-k+1; i++ {
		temp = append(temp, i)
		combineTrack(temp, combinations, i+1, k-1, n)
		temp = temp[:len(temp)-1]
	}
}
